### ISSUE TITLE
CSS styling doesn't work in Handlebars chart

### ISSUE THREAD

**Issue by goldjee** [apache/superset#30381]
### Bug description

CSS code defined in "CSS styles" section of Handlebars chart settings isn't applied to HTML template. It is shown as plain text on the chart instead.

### How to reproduce the bug

1. Disable ESCAPE_MARKDOWN_HTML, enable HTML_SANITIZATION, allow CSS with HTML_SANITIZATION_SCHEMA_EXTENSIONS in `superset_config.py` as follows:
```Python
FEATURE_FLAGS = {
    "ESCAPE_MARKDOWN_HTML": False,
    "HTML_SANITIZATION": True,
}

HTML_SANITIZATION_SCHEMA_EXTENSIONS = {
    "attributes": {
    "*": ["style","className"],
    },
    "tagNames": ["style"],
}
```
2. Create a Handlebars chart with one column ("descr" in this example) added from any dataset.
3. Set up HTML template as follows:
```HTML
<ul class="data-list">
  {{#each data}}
    <li>{{descr}}</li>
  {{/each}}
</ul>
```
5. Set up CSS for the chart:
```CSS
.data-list {
    background-color: yellow;
}
```
7. Observe your CSS code in the preview as a plain text.

### Screenshots/recordings

<img width="1101" alt="image" src="https://github.com/user-attachments/assets/271fa5da-656b-4571-bf38-e0fbcd266f76">

### Superset version

master / latest-dev

### Python version

3.11

### Node version

18 or greater

### Browser

Chrome

### Additional context

While investigating this issue, I dived into Handlebars code. There is a section in `superset-frontend/plugins/plugin-chart-handlebars/src/components/Handlebars/HandlebarsViewer.tsx` at lines 35-43 that seemingly retrieves HTML sanitization and schema overrides. I went to the browser console and printed these variables with this snippet:
```JS
const appContainer = document.getElementById('app');
const { common } = JSON.parse(
    appContainer?.getAttribute('data-bootstrap') || '{}',
);
const htmlSanitization = common?.conf?.HTML_SANITIZATION;
const htmlSchemaOverrides =
    common?.conf?.HTML_SANITIZATION_SCHEMA_EXTENSIONS;

console.log(htmlSanitization);
console.log(htmlSchemaOverrides);
```
The output returns true and empty object respectively despite the overrides are configured (see "How to reproduce the bug" section). It appears that either I'm doing something wrong, or the overrides are not read from the `superset_config.py`.

### Checklist

- [x] I have searched Superset docs and Slack and didn't find a solution to my problem.
- [x] I have searched the GitHub issue tracker and didn't find a similar bug report.
- [x] I have checked Superset's logs for errors and if I found a relevant Python stacktrace, I included it here as text in the "additional context" section.

**Comment by goldjee** [comment]
@dosu
Yep, I've seen that similar issue and tried the steps described there. HTML is parsed just fine in my case, it's just CSS malfunctioning.

I've accidentally posted the issue too early and edited in some additional info in the "Additional context" section that may be useful in this case. Despite `HTML_SANITIZATION_SCHEMA_EXTENSIONS` is configured, it seems that this parameter is not taken into account.

If I disable HTML sanitization completely, the CSS is not applied as well.

**Comment by goldjee** [comment]
@dosu
Can you provide an example of full superset_config.py with sanitization configured?

**Comment by goldjee** [comment]
@dosu
Oh I see, `HTML_SANITIZATION` is not a feature flag, but a top level variable. I've fixed that in my config and double checked `HTML_SANITIZATION_SCHEMA_EXTENSIONS`, but the issue is still there.

Where in the project codebase can I find the bridge that passes `HTML_SANITIZATION_SCHEMA_EXTENSIONS` from the Python backend to the frontend? I'd like to try to check if the parameter is actually passed.

**Comment by goldjee** [comment]
@dosu
Which file in the Superset codebase defines Common Bootstrap Payload? Could you provide the path to it?

**Comment by goldjee** [comment]
@dosu
There was a mistake on my side: I was passing wrong config file while starting Superset. Fixed that. Now I can see that `HTML_SANITIZATION_SCHEMA_EXTENSIONS` is sent to the frontend correctly.

Anyway, when I pass the config file with
```Python
HTML_SANITIZATION = True
HTML_SANITIZATION_SCHEMA_EXTENSIONS = {
    "attributes": {
    "*": ["style","className"],
    },
    "tagNames": ["style"],
}
```
the `class` attribute is blank for processed HTML template.

My chart config is as follows. HTML:
```HTML
<ul class="data-list">
  {{#each data}}
    <li>{{descr}}</li>
  {{/each}}
</ul>
```
CSS:
```CSS
.data-list {
    background-color: yellow;
    list-style: none;
}
```
Output HTML in the chart is:
```HTML
<ul class="">
    <li>Quisque vehicula pharetra velit, vel pharetra erat pellentesque a. Maecenas molestie maximus sem, ac dapibus.</li>
    <li>Aliquam posuere malesuada pharetra. Pellentesque id vulputate elit. Proin tristique fringilla odio vel vehicula. Cras.</li>
</ul>
```
If I disable sanitization, the CSS is applied properly. It points out that my `HTML_SANITIZATION_SCHEMA_EXTENSIONS` is probably missing something.

**Comment by goldjee** [comment]
@dosu
I've included `class` to allowed attribute list and restarted Superset. Updated config looks like this:
```Python
HTML_SANITIZATION = True
HTML_SANITIZATION_SCHEMA_EXTENSIONS = {
    "attributes": {
    "*": ["style","className","class"],
    },
    "tagNames": ["style"],
}
```
The `class` attribute in the output HTML is still empty.

**Comment by goldjee** [comment]
@dosu
`className` was there because it was present in other issues and in official article regarding Handlebars security in Superset [here](https://preset.io/blog/superset-patch-release/#handlebars-chart-configuration). Anyway, I've tried excluding `className` and this trick didn't work.

**Comment by goldjee** [comment]
@dosu
Should `xss` library be configured somewhere in config files? I've checked `superset-frontend/packages/superset-ui-core/src/utils/html.tsx` and its contents match with the snippet you have provided. `class` attribute is there in the whitelist.

**Comment by goldjee** [comment]
Seems like at this moment I've hit the wall with the bot. I would really appreciate if the community could help me to proceed with this issue.

**Comment by sfirke** [comment]
Hi!  You got somewhere with the bot at least, I was gonna say `HTML_SANITIZATION = True` is in my top level superset_config.py, not a Feature Flag.  But looks like you got that.

I have this, which includes some options for iFraming you might not need.  I think this is nothing new but just to share in case:
```
HTML_SANITIZATION = True
HTML_SANITIZATION_SCHEMA_EXTENSIONS = {
  "attributes": {
    "*": ["style", "className", "src", "width", "height", "frameborder", "marginwidth", "marginheight", "scrolling", "target"], # this enables basic CSS styling for Handlebars text charts and iFrame embeds for our feedback forms
  },
  "tagNames": ["style", "iframe", "h1", "h2", "h3", "h4", "h5", "h6"],
}
```

You might need to enable `unsafe-eval` in your Talisman CSP, see https://github.com/apache/superset/issues/27299#issuecomment-1978075353 and https://github.com/apache/superset/issues/25205 for instance

**Comment by goldjee** [comment]
Hi, @sfirke! I've modified `HTML_SANITIZATION_SCHEMA_EXTENSIONS` and added Talisman config like this:
```Python
HTML_SANITIZATION = True
HTML_SANITIZATION_SCHEMA_EXTENSIONS = {
  "attributes": {
    "*": ["style", "className", "class", "src", "width", "height", "frameborder", "marginwidth", "marginheight", "scrolling", "target"],
  },
  "tagNames": ["style", "iframe", "h1", "h2", "h3", "h4", "h5", "h6"],
}

TALISMAN_ENABLED = True
TALISMAN_CONFIG = {
    "content_security_policy": {
        "default-src": ["'self'", "'unsafe-inline'", "'unsafe-eval'"]
    }
}
```

I can see updated schema extensions at the client side:
<img width="519" alt="image" src="https://github.com/user-attachments/assets/97f0dcea-b312-4fbb-8c24-d0d7cc784e98">

However, these adjustments didn't do the trick for me. The `class` attribute is still empty.

<img width="1066" alt="image" src="https://github.com/user-attachments/assets/15158ea0-480a-4b4e-aa9f-e492b3084321">

I've also noticed an error in the browser console:
```
Uncaught NetworkError: Failed to execute 'importScripts' on 'WorkerGlobalScope': The script at 'http://localhost:8088/static/assets/worker-css.js' failed to load.
    at 408b31b0-bd74-41d2-bcda-02d1e74d554e:1:1
```
Wonder if it has something to do with the missing attribute.

**Comment by sfirke** [comment]
If you are just working locally for now, can you try setting Talisman enabled to False? Then that's one thing we can remove from consideration if that doesn't fix it.

**Comment by sfirke** [comment]
Also I see you are working on master / latest. I am working on 4.0.2 the latest official release. Maybe a bug has been introduced since then? Can you try 4.0.2?

**Comment by goldjee** [comment]
Yep, it's local installation based on master without Docker. Disabled Talisman, `class` is still empty and the same error is present in the console.

Switched to `4.0` branch, rebuilt the app (both frontend and backend) and restarted the service. Same outcome :(

**Comment by goldjee** [comment]
My colleague has managed to set everything up at our corporate instance, which uses 4.0.2. The config is pretty much the same, so it's either something wrong with my local installation, or with `master`. I lean towards the former.

**Comment by sfirke** [comment]
Well I'm glad it's working elsewhere for your team.  One more possibility for something that could be wrong with a local installation: I saw someone mention in a comment recently (can't find it again) that they had to turn off a flag like DEBUG_MODE or similar, that when it was true it was overriding their Talisman settings.  Though when I look [at the code](https://github.com/apache/superset/blob/0fdcd8b27e6d6ba70f8f38b4a033c61555af92bd/superset/initialization/__init__.py#L621), it looks like if you set `TALISMAN_ENABLED = False` that would take care of things.

**Comment by goldjee** [comment]
Thanks for your help, @sfirke !

I couldn't find a flag similar to `DEBUG_MODE` in the documentation. But as CSS styling actually works at 4.0.2 production and this issue is mostly caused by my own local setup, I think it will be easier to just disable `HTML_SANITIZATION` for now.

**Comment by mahesh-babu-drivex** [comment]
> Thanks for your help, @sfirke !
> 
> I couldn't find a flag similar to `DEBUG_MODE` in the documentation. But as CSS styling actually works at 4.0.2 production and this issue is mostly caused by my own local setup, I think it will be easier to just disable `HTML_SANITIZATION` for now.

@goldjee Hey goldjee can you tell me what worked for you what is the config you followed to make it worked?

**Comment by goldjee** [comment]
@mahesh-babu-drivex 

These are related settings in my local config:
```Python
FEATURE_FLAGS = {
    'ENABLE_TEMPLATE_PROCESSING': True,
    'ESCAPE_MARKDOWN_HTML': False
}

HTML_SANITIZATION = False
HTML_SANITIZATION_SCHEMA_EXTENSIONS = {
  "attributes": {
    "*": ["style", "className", "class"],
  },
  "tagNames": ["style"],
}

TALISMAN_ENABLED = False
TALISMAN_CONFIG = {
    "content_security_policy": {
        "default-src": ["'self'", "'unsafe-inline'", "'unsafe-eval'"]
    }
}
```

Works fine for my purposes, but probably is not safe to use in production.

**Comment by sfirke** [comment]
This looks like the same bug as #43721, with a different cause than the one diagnosed here.

It isn't `HTML_SANITIZATION_SCHEMA_EXTENSIONS` — the split happens in `remark-parse`, before `rehype-sanitize` ever runs. `Handlebars.tsx` joins the template and the `<style>` block with a single newline, so when the template opens with a block-level HTML tag the `<style>` tag gets absorbed into that tag's Markdown HTML block (type 6), which ends at the first blank line — leaking every CSS rule after it as visible text.

Workaround without a patch: add a trailing blank line at the end of the Handlebars Template box, or remove all blank lines from the CSS.

Fix in #43722.
